### PR TITLE
Update details.js DOM text reinterpreted as HTML

### LINKED ISSUE #38766: [Issue] Update details.js DOM text reinterpreted as HTML
This issue is automatically created based on existing pull request: magento/magento2#38757: Update details.js DOM text reinterpreted as HTML

---------

### Description (*)
By using textContent , it will avoid the risk of HTML injection, as these properties automatically escape any HTML special characters in the provided text. This helps prevent cross-site scripting (XSS) vulnerabilities by treating the input as plain text rather than interpreted HTML.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by Shivam7-1**
### Description (*)
By using textContent , it will avoid the risk of HTML injection, as these properties automatically escape any HTML special characters in the provided text. This helps prevent cross-site scripting (XSS) vulnerabilities by treating the input as plain text rather than interpreted HTML.

### To help with the testing you can use the following manual testing scenarios:

1. **Check if HTML tags are correctly escaped:**  
   Add a product with various HTML tags (e.g., `<b>bold</b>`, `<script>alert("test")</script>`) in the cart, then verify that these are displayed correctly without being interpreted as HTML on the checkout page.

2. **Check edge cases with special characters:**  
   Add products with special characters in their names (e.g., `&`, `<`, `>`, `"`). Ensure that these characters are correctly sanitized and displayed as expected.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#38766: Update details.js DOM text reinterpreted as HTML

### CODE REVIEW

**Comment by Shivam7-1** [comment]
@magento run all tests

**Comment by Shivam7-1** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento create issue

**Comment by Shivam7-1** [comment]
Hii @engcom-Charlie and @engcom-Hotel Could You Please Review This PR 
As Soon As Possible Because its pending for Review since many months 

Thanks & Regards

**Comment by Shivam7-1** [comment]
Hii @engcom-Charlie , @engcom-Dash @engcom-Bravo and @engcom-Hotel Could You Please Review This PR
As Soon As Possible Because its pending for Review since many months

Thanks & Regards

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/Checkout/view/frontend/web/js/view/summary/item/details.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ define([
         getNameUnsanitizedHtml: function (quoteItem) {
             var txt = document.createElement('textarea');
 
-            txt.innerHTML = quoteItem.name;
```
Comment (engcom-Hotel) [inline app/code/Magento/Checkout/view/frontend/web/js/view/summary/item/details.js:25]: Hello @Shivam7-1,

Please correct me if I am wrong, but here we are initializing the `txt.innerHTML` and in the next line the same has been sanitized using `escaper.escapeHtml(txt.value, this.allowedTags)`. Does this change is even required?

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Shivam7-1,

Please refer to the below review comment.

Thanks

File: app/code/Magento/Checkout/view/frontend/web/js/view/summary/item/details.js (hunk 1 above)
Comment (Shivam7-1) [reply app/code/Magento/Checkout/view/frontend/web/js/view/summary/item/details.js:25]: Hii @engcom-Hotel Thanks For Review But I think yes Using textContent is more beneficial here as it automatically ensures text is safely handled without risking HTML injection, unlike innerHTML.

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel  Thanks For Review But I think yes Using textContent is more beneficial here as it automatically ensures text is safely handled without risking HTML injection, unlike innerHTML.

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel  Could You Please Review This again 
Thanks

File: app/code/Magento/Checkout/view/frontend/web/js/view/summary/item/details.js (hunk 1 above)
Comment (engcom-Hotel) [reply app/code/Magento/Checkout/view/frontend/web/js/view/summary/item/details.js:25]: Yes @Shivam7-1, I agree that using `textContent` is beneficial in order to prevent the HTML injection, but for the same we are already sanitizing the inputs here:

https://github.com/magento/magento2/blob/5184c0675bdec00396e2e555344d85d133d74de6/app/code/Magento/Checkout/view/frontend/web/js/view/summary/item/details.js#L27

Or you can provide us the manual testing scenario to test this PR change, it will help us with proceed further with this PR.

**Comment by engcom-Hotel** [comment]
Hello @Shivam7-1,

Please refer to our review comment here https://github.com/magento/magento2/pull/38757#discussion_r1877447360.

Thanks

**Comment by Shivam7-1** [comment]
Hi @engcom-Hotel 
Thanks for the feedback. I understand that the code already sanitizes the input using `escaper.escapeHtml`. The reason for using `textContent` in the updated code is to prevent any potential HTML injection before even starting the sanitization process, as it ensures that the value is treated purely as text and not interpreted as HTML.

While I agree that the sanitization step exists, I believe using `textContent` helps mitigate any risks at an earlier stage, especially in cases where there might be unexpected characters or encodings. 

To help with the testing you can use the following manual testing scenarios:

1. **Check if HTML tags are correctly escaped:**  
   Add a product with various HTML tags (e.g., `<b>bold</b>`, `<script>alert("test")</script>`) in the cart, then verify that these are displayed correctly without being interpreted as HTML on the checkout page.

2. **Check edge cases with special characters:**  
   Add products with special characters in their names (e.g., `&`, `<`, `>`, `"`). Ensure that these characters are correctly sanitized and displayed as expected.

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel Could You Please Review This again
Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel is there Anything Else is Required From My side to get this PR merge
Thanks

**Comment by engcom-Hotel** [comment]
Hello @Shivam7-1,

We have tried to reproduce the issue with the given manual testing scenarios with these PR changes but it seems the issue is not fixed.

We have created the below products and added them to the cart:
image

And then we checked in the front on the checkout page, we could see the product name below:

image

Please recheck the changes and let us know if we missed anything.

Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel Thanks For Reviewing 
You can reply with something like this:

The purpose of replacing `innerHTML` with `textContent` is to prevent any HTML tags or scripts from being executed, effectively mitigating XSS risks. This ensures that any potentially dangerous content, such as `<script>` tags or event handlers (like `onerror`), is not rendered in the browser.

However, based on your testing, it seems some product names are still being impacted by how they're displayed (for example, the `<b>` tag still shows bold text). This behavior is expected, as `textContent` will strip out any HTML formatting, which may not be desirable in some use cases, such as for product names where formatting is needed.

If this is the case, and the text content must allow HTML for formatting purposes, I recommend we explore using a solution like `DOMPurify` to sanitize and allow only safe HTML. This would provide an additional layer of protection while preserving the necessary HTML formatting.

Please let me know how you'd like to proceed, and if moving forward with **DOMPurify** would be a viable solution for the project.

Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel Could you please Review this PR again 
Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel Could you please Review this PR again
Thanks

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Hotel** [comment]
Hello @Shivam7-1,

At this stage, no action is required from your side. The PR is currently in the testing bucket and will be progressed based on priority. We appreciate your patience.

Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel Thanks For Reviewing If possible Could Team Fastrack the Process 
Because this PR is created long back ago 
Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel  Could you Please ping tester here for tasting and get merge this PR 

So it will get merge Soon 
Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel And this also 
Is there any Update on this PR Testing?
Or could you also please assign anyone who Could test this ?

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel Could you Assign or ping anyone from Team to Test this PR 
Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Hotel Just Follow Up on this PR
Thanks

**Comment by engcom-Bravo** [comment]
Hi @Shivam7-1,

Thanks for the collaboration & contribution!

### ❌ : QA FAILED


**Before:** :heavy_multiplication_x:


Screenshot 2025-01-17 at 12 44 35 pm



**:x: Actual result After Fix** 




Screenshot 2025-01-17 at 12 44 35 pm




The results were same even after taking the PR changes.Could you Please let us know if we are missing anything and elaborate the expected and actual results.


Thanks.

**Comment by Shivam7-1** [comment]
Hii @engcom-Bravo  Thanks For Reviewing  above it's working fine you can check this comments here https://github.com/magento/magento2/pull/38757#issuecomment-2535287874
https://github.com/magento/magento2/pull/38757#issuecomment-2535312354

**Comment by Shivam7-1** [comment]
Hii @engcom-Bravo Thanks For Testing Anything else is Required from my side 
Thanks

**Comment by engcom-Bravo** [comment]
Hi @Shivam7-1,

Thanks for your Update.

Could you please let us know the aim of the PR and the issue that is solving here with the help of this PR.As per this comment https://github.com/magento/magento2/pull/38757#issuecomment-2598111623 before and after PR changes there is no change as such.

Kindly provide us detailed manual testing scenario to proceed further.

Thanks.

**Comment by Shivam7-1** [comment]
Hii @engcom-Bravo 
Thank you for your feedback!

The aim of this PR is to mitigate the potential risk of HTML injection, particularly related to XSS (Cross-Site Scripting) vulnerabilities. In the original code, we were using `txt.innerHTML = quoteItem.name;`, which could allow malicious HTML content to be injected into the DOM, depending on the input. This could lead to vulnerabilities if the input contains malicious scripts.

By switching to `txt.textContent = quoteItem.name;`, we ensure that any content is treated as plain text. This approach automatically escapes any special HTML characters and prevents malicious HTML from being injected. However, I’ve kept the `escaper.escapeHtml` function in place to handle any further sanitization related to allowed tags.

Regarding the manual testing, here are the scenarios which could be tested to ensure functionality:

1. **Safe Input Test**: Provided regular text in `quoteItem.name` to confirm that it renders correctly as plain text.
2. **XSS Injection Test**: Attempted to inject HTML/JS (e.g., `<script>alert('XSS');</script>`) in the `quoteItem.name` field and verified that the code doesn't execute and is rendered as plain text.
3. **Special Characters Test**: Used special characters such as `<`, `>`, `&`, etc., in the input to verify they are properly escaped and displayed as intended.

The changes mainly affect the way the input is handled in terms of security. While it might seem like there’s no visible change, this update significantly improves the security posture by preventing potential XSS attacks.

Please let me know if you need further clarifications or additional tests!

Thanks,  
Shivam

**Comment by Shivam7-1** [comment]
Hii @engcom-Bravo  anything else is required from my side for this PR

**Comment by Shivam7-1** [comment]
Hii  @engcom-Bravo and @engcom-Hotel Could anyone team merge this PR if there is no changes is required 
Thanks

**Comment by Shivam7-1** [comment]
Hii @engcom-Charlie , @engcom-Dash @engcom-Bravo and @engcom-Hotel Team now this PR is ready to merge i think since 1.5 month could team look into this 

Thanks

**Comment by engcom-Charlie** [comment]
Hi @Shivam7-1,

Currently, no action is required from your side. The PR is currently in the testing bucket and team will pick this PR for further processing based on priority. Thank you for your patience.

**Comment by Shivam7-1** [comment]
Hii @engcom-Charlie 
Is there any update on this PR merging 

Thanks

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
Hello @Shivam7-1 

Thanks for your PR addressing XSS concerns. I wanted to share some findings from my testing:

To reproduce the issue and validate the fix, I temporarily removed the escaper.escapeHtml logic in the function. However, even without the escaper, Content Security Policy (CSP) is still blocking the output when it's bound using Knockout's html: binding. This is because CSP restricts inline scripts or HTML that could potentially execute malicious JavaScript

image

image


image

If we take a look at app/code/Magento/Security/view/base/web/js/escaper.js "**escapeHtml**" function is specifically designed to escape a string for safe injection into HTML, converting characters like <, >, &, and " into their respective HTML entities. This ensures that any potentially dangerous content in quoteItem.name is rendered as plain text in the browser, rather than being interpreted as HTML or JavaScript.

please find below text where script rendered as plain text:

Screenshot 2025-06-20 at 3 26 20 PM

In this case, the content is not being rendered directly it's being escaped first. So the change doesn’t add any additional security benefit.
